File: widget/app/src/main/java/com/example/structuredwidget/data/TodayRepository.kt
package com.example.structuredwidget.data

import java.time.LocalDate
import java.time.LocalTime
import java.time.ZoneId

class TodayRepository(
    private val nowProvider: () -> LocalTime = { LocalTime.now() },
) {
    /** Build today state from a single atomic snapshot. */
    fun fromSnapshot(snapshot: WidgetSnapshot): TodayState {
        val today = snapshot.logicalDate
        val now = nowProvider()
        val nowHours = now.hour + now.minute / 60.0

        val inbox = snapshot.inbox.filter { it.completedAt.isNullOrBlank() }
        val due = snapshot.due
            .filter { it.completedAt.isNullOrBlank() }
            .sortedWith(
                compareByDescending<StructuredTask> { it.day.orEmpty() }
                    .thenBy { it.startTime ?: Double.MAX_VALUE },
            )
        val tomorrow = snapshot.tomorrow.filter { it.completedAt.isNullOrBlank() }

        val all = snapshot.today
            .filter { it.completedAt.isNullOrBlank() }

        val sorted = all.sortedWith(
            compareBy<StructuredTask> { it.isAllDay }
                .thenBy { it.startTime ?: Double.MAX_VALUE },
        )
        val allDay = sorted.filter { it.isAllDay }
        val timed = sorted.filter { !it.isAllDay }

        val inProgress = timed.firstOrNull { t ->
            val start = t.startTime ?: return@firstOrNull false
            val end = start + t.duration / 60.0
            nowHours in start..end
        }
        val upcoming = timed.filter {
            val start = it.startTime ?: return@filter false
            start > nowHours
        }
        val hero = inProgress ?: upcoming.firstOrNull() ?: timed.lastOrNull()

        val upNext = timed
            .filter { it.id != hero?.id }
            .filter { t ->
                val start = t.startTime
                start == null || start >= nowHours - 0.5
            }
            .take(6)

        val accent = try {
            hero?.color?.let { android.graphics.Color.parseColor(it) }
        } catch (_: Exception) {
            null
        }

        return TodayState(
            allDay = allDay,
            hero = hero,
            upNext = upNext,
            inbox = inbox,
            tomorrow = tomorrow,
            due = due,
            currentAccent = accent,
            isEmpty = all.isEmpty() && inbox.isEmpty() && due.isEmpty(),
            logicalDate = today,
            serverTimezone = snapshot.timezone,
            timezoneMismatch = isTimezoneMismatch(snapshot.timezone),
        )
    }

    /** Legacy multi-fetch path for tests using [StructuredTaskSource]. */
    suspend fun load(source: StructuredTaskSource, clock: () -> LocalDate = { LocalDate.now() }): TodayState {
        val today = clock()
        val now = nowProvider()
        val nowHours = now.hour + now.minute / 60.0

        val oneOff = source.fetchOneOffTasks(today)
        val recurring = source.fetchRecurringTasks()
        val inbox = source.fetchInboxTasks().filter { it.completedAt.isNullOrBlank() }
        val due = source.fetchOpenTasks()
            .filter { it.completedAt.isNullOrBlank() }
            .sortedWith(
                compareByDescending<StructuredTask> { it.day.orEmpty() }
                    .thenBy { it.startTime ?: Double.MAX_VALUE },
            )
        val tomorrow = source.fetchForwardTasks(today.plusDays(1), today.plusDays(1))
            .filter { it.completedAt.isNullOrBlank() }

        val all = (oneOff + recurring)
            .filter { it.day == null || it.day == today.toString() }
            .filter { it.completedAt.isNullOrBlank() }

        val sorted = all.sortedWith(
            compareBy<StructuredTask> { it.isAllDay }
                .thenBy { it.startTime ?: Double.MAX_VALUE },
        )
        val allDay = sorted.filter { it.isAllDay }
        val timed = sorted.filter { !it.isAllDay }

        val inProgress = timed.firstOrNull { t ->
            val start = t.startTime ?: return@firstOrNull false
            val end = start + t.duration / 60.0
            nowHours in start..end
        }
        val upcoming = timed.filter {
            val start = it.startTime ?: return@filter false
            start > nowHours
        }
        val hero = inProgress ?: upcoming.firstOrNull() ?: timed.lastOrNull()

        val upNext = timed
            .filter { it.id != hero?.id }
            .filter { t ->
                val start = t.startTime
                start == null || start >= nowHours - 0.5
            }
            .take(6)

        val accent = try {
            hero?.color?.let { android.graphics.Color.parseColor(it) }
        } catch (_: Exception) {
            null
        }

        return TodayState(
            allDay = allDay,
            hero = hero,
            upNext = upNext,
            inbox = inbox,
            tomorrow = tomorrow,
            due = due,
            currentAccent = accent,
            isEmpty = all.isEmpty() && inbox.isEmpty() && due.isEmpty(),
            logicalDate = today,
            serverTimezone = null,
            timezoneMismatch = false,
        )
    }

    companion object {
        fun isTimezoneMismatch(serverTimezone: String): Boolean {
            if (serverTimezone.isBlank()) return false
            val deviceTz = ZoneId.systemDefault().id
            return !deviceTz.equals(serverTimezone, ignoreCase = true)
        }
    }
}

File: ios/Sources/StructuredCore/TodayRepository.swift
import Foundation

public struct TodayRepository: Sendable {
    public var now: @Sendable () -> ClockTime
    public var deviceTimeZoneId: @Sendable () -> String

    public init(
        now: @escaping @Sendable () -> ClockTime = { ClockTime.now() },
        deviceTimeZoneId: @escaping @Sendable () -> String = { TimeZone.current.identifier }
    ) {
        self.now = now
        self.deviceTimeZoneId = deviceTimeZoneId
    }

    public func fromSnapshot(_ snapshot: WidgetSnapshot) -> TodayState {
        let today = snapshot.logicalDate
        let nowHours = now().hours

        let inbox = snapshot.inbox.filter { !$0.isCompleted }
        let due = snapshot.due
            .filter { !$0.isCompleted }
            .sorted { lhs, rhs in
                let ld = lhs.day ?? ""
                let rd = rhs.day ?? ""
                if ld != rd { return ld > rd }
                let ls = lhs.startTime ?? Double.greatestFiniteMagnitude
                let rs = rhs.startTime ?? Double.greatestFiniteMagnitude
                return ls < rs
            }
        let tomorrow = snapshot.tomorrow.filter { !$0.isCompleted }

        let all = snapshot.today.filter { !$0.isCompleted }
        let sorted = all.sorted { lhs, rhs in
            if lhs.isAllDay != rhs.isAllDay { return lhs.isAllDay && !rhs.isAllDay }
            let ls = lhs.startTime ?? Double.greatestFiniteMagnitude
            let rs = rhs.startTime ?? Double.greatestFiniteMagnitude
            return ls < rs
        }
        let allDay = sorted.filter(\.isAllDay)
        let timed = sorted.filter { !$0.isAllDay }

        let inProgress = timed.first { t in
            guard let start = t.startTime else { return false }
            let end = start + Double(t.duration) / 60.0
            return nowHours >= start && nowHours <= end
        }
        let upcoming = timed.filter { t in
            guard let start = t.startTime else { return false }
            return start > nowHours
        }
        let hero = inProgress ?? upcoming.first ?? timed.last

        let upNext = timed
            .filter { $0.id != hero?.id }
            .filter { t in
                guard let start = t.startTime else { return true }
                return start >= nowHours - 0.5
            }
            .prefix(6)

        return TodayState(
            allDay: allDay,
            hero: hero,
            upNext: Array(upNext),
            inbox: inbox,
            tomorrow: tomorrow,
            due: due,
            currentAccent: hero?.color,
            isEmpty: all.isEmpty && inbox.isEmpty && due.isEmpty,
            logicalDate: today,
            serverTimezone: snapshot.timezone,
            timezoneMismatch: Self.isTimezoneMismatch(snapshot.timezone, device: deviceTimeZoneId())
        )
    }

    public static func isTimezoneMismatch(_ serverTimezone: String, device: String) -> Bool {
        if serverTimezone.isEmpty { return false }
        return serverTimezone.caseInsensitiveCompare(device) != .orderedSame
    }
}

public struct WeekRepository: Sendable {
    public init() {}

    public func fromSnapshot(_ snapshot: WidgetSnapshot) -> WeekState {
        let today = snapshot.logicalDate
        let tasks = snapshot.week
        let days: [DayBlock] = (0...7).map { offset in
            let date = today.adding(days: offset)
            let dayTasks = tasks.filter { $0.day == date.isoString }
            let sorted = dayTasks.sorted { ($0.startTime ?? .greatestFiniteMagnitude) < ($1.startTime ?? .greatestFiniteMagnitude) }
            let timed = sorted.filter { !$0.isAllDay }
            let allDay = sorted.filter(\.isAllDay)
            return DayBlock(
                date: date,
                timed: timed,
                allDay: allDay,
                totalTasks: timed.count + allDay.count,
                isToday: date == today,
                isYesterday: date == today.adding(days: -1)
            )
        }
        return WeekState(days: days)
    }
}

public enum WidgetHeader {
    public static func todayTaskCount(_ todayState: TodayState?) -> Int {
        guard let todayState else { return 0 }
        let heroExtra = todayState.hero == nil ? 0 : 1
        return todayState.allDay.count + todayState.upNext.count + heroExtra
    }
}
